### DISCUSSION TITLE
Is there a way to ignore changes made by a given field manager, similar to ignoreDifferences in argo?

### DISCUSSION THREAD

**Discussion by mccomplete** [fluxcd/flux2#4925] [General]
We have an agent that dynamically changes fields of deployments running in the cluster. Is there any way to tell flux to ignore changes made by this agent, similar to `resource.customizations.ignoreDifferences.all` with `managedFieldManagers` in argo cd?

**Comment by kingdonb** [comment]
There's a few differences between what you suggested and what's available in Flux, but yes you can, sort of.

The default mode is designed to override any differences because that's the core expectation of the majority of Flux users. There is an `ssa: merge` annotation that you can use to override that expectation. Then, Flux will allow divergences in unspecified parts of the config. If your definition conflicts with the definition applied by another field manager, then it will flap between whatever state you specified in Git and the field manager state. But if your git configuration is able to merge cleanly with the fields added by field manager, it should work as you expect.

There are a few more details I can provide when I can get back to the reference material, but why don't you tell us a bit about your use case and if that solves your issue before I go into any more details? I don't want to give too much irrelevant information.

There are a number of different special cases like the blessed list of field managers, and common issues where there would be a conflict but isn't because of how Flux worked around them. (Like with persistent volumes managed in Git for example.)

**Comment by mccomplete** [comment]
Here is an example: say there is a deployment in the cluster called "hello" with one container. In the git repo, the "hello" container is defined as requesting 500m CPU. Our agent changes the CPU request to 50m. We want flux to leave the deployment as it is, even though it diverges from the definition in git.

On the other hand, we want flux to update the deployment if there are any other changes, e.g. the container image or env vars or anything else. In that case, our admission controller will preserve our changes. We just don't want flux to be trying to change the deployment every time flux reconciles the cluster (e.g. every 5 minutes).

In an environment with argo, we can solve this problem by setting a property `resource.customizations.ignoreDifferences.all` with `managedFieldManagers` and the name of our field manager.

Can we accomplish something like this with the `ssa: merge` annotation? If so, how?

**Comment by stefanprodan** [comment]
Currently only the fields that are mutated at admission are left in place.

**Comment by mccomplete** [comment]
@stefanprodan  Not sure what you mean. Do you mean that any field that is mutated by an admission controller is not considered a divergence by flux the next time it tries to reconcile the cluster? 

Or are you saying that flux will try to revert the deployment every time it reconciles (e.g. every five minutes), and our admission controller will have to preserve our changes?

**Reply by kingdonb** [reply]
The outcome of a dry-run server side includes any mutating effects from admission controllers. If the admission controller changes the Git `500m` into `50m` then Flux will not see any diff to correct, and should honor the admission controller's setting.

**Comment by geoshapka** [comment]
Hi, adding to discussion:
In HelmRelease we have driftDetection.ignore, that does pretty much what we need.
Can similar functionality be added to Kustomization in some form?

**Comment by gera-corp** [comment]
Hi, I'm having a issue with mutations kind like Secret, ConfigMap, and custom CRDs. It would be great if we had some kind of mechanism, for example, ignoreDifferences.

**Comment by srkn0** [comment]
+1. Running into the same issue with Cluster API controllers that modify resources post-deployment. Having something like ArgoCD's ignoreDifferences would be really helpful for these cases.
